### PR TITLE
Batch Write API for collections and String

### PR DESCRIPTION
**Pull request by ZiyanShi**
Signed-off-by: ZiyanShi <ziyan.shi@intel.com>

### What problem does this PR solve?

Add batch write API for collections and String.
Reimplement batch write for string to make code compatible with collections.

### Tests

Unit test

### Performance

Batch Write performance for String has no noticeable change.

### CODE REVIEW

File: engine/write_batch_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#include "write_batch_impl.hpp"
+
+#include "alias.hpp"
+
+namespace KVDK_NAMESPACE {
+
+std::string BatchWriteLog::Serialize() {
```
Comment (JiayuZzz) [inline engine/write_batch_impl.cpp:7]: Maybe Encode/Decode for alignment

Comment (JiayuZzz) [inline engine/write_batch_impl.cpp:7]: Maybe Encode/Decode for alignment

Comment (ZiyanShi) [reply engine/write_batch_impl.cpp]: done

File: engine/kv_engine.cpp [hunk 2]
```diff
@@ -740,130 +710,303 @@ Status KVEngine::CheckConfigs(const Configs& configs) {
...
+  for (auto const& string_op : batch.StringOps()) {
+    StringWriteArgs args;
+    args.Assign(string_op);
+    args.ts = bw_token.Timestamp();
+    args.res = lookupKey<true>(args.key, StringRecordType);
+    if (args.res.s != Status::Ok && args.res.s != Status::NotFound &&
+        args.res.s != Status::Outdated) {
+      return args.res.s;
+    }
+    if (args.op == WriteBatchImpl::Op::Delete && args.res.s != Status::Ok) {
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:803]: res.s might be memory overflow

**Review by JiayuZzz — APPROVED** [review]
LGTM

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -740,130 +710,303 @@ Status KVEngine::CheckConfigs(const Configs& configs) {
...
+      // No need to do anything for delete a non-existing String
+      continue;
+    }
+    args.space = pmem_allocator_->Allocate(
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:807]: for delete, make sure string_op.value is empty? or if(delete) { RecordSize(key, ""); }

File: engine/write_batch_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+  void Assign(WriteBatchImpl::StringOp const& string_op) {
+    key = string_op.key;
+    value = string_op.value;
+    op = string_op.op;
+  }
+};
+
+struct SortedWriteArgs {
+  StringView key;
```
Comment (sherry-1001) [inline engine/write_batch_impl.hpp:105]: maybe better use a common struct.

File: engine/write_batch_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+  size_t Size() const {
+    return string_logs.size() + sorted_logs.size() + hash_logs.size();
+  }
+
+  static size_t Capacity() { return (1UL << 20); }
+
+  static size_t MaxBytes() {
+    return sizeof(size_t) + sizeof(TimeStampType) + sizeof(Stage) +
+           sizeof(size_t) + Capacity() * sizeof(StringLog);
```
Comment (sherry-1001) [inline engine/write_batch_impl.hpp:214]: why not `+sizeof(SortedLog)` `+sizeof(HashLog)`

File: tests/tests.cpp [hunk 6]
```diff
@@ -1983,191 +1991,64 @@ TEST_F(EngineBasicTest, TestbackgroundDestroyCollections) {
...
-    // thread 4
-    ts.emplace_back([&]() {
-      TEST_SYNC_POINT("Test::Delete::Start::0");
-      for (int i = 0; i < batch_size; i += 3) {
-        std::string key = "key" + std::to_string(i);
-        ASSERT_EQ(engine->Delete(key), Status::Ok);
-      }
-    });
+TEST_F(EngineBasicTest, BatchWriteRollBack) {
+  size_t num_threads = 16;
```
Comment (sherry-1001) [inline tests/tests.cpp:1996]: please add comments for this case.

**Review by sherry-1001 — APPROVED** [review]

**Pull request merged**